src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JRTArchive.java

```diff
@@ -0,0 +1,536 @@
+                                                   "--patch-module is not supported for run-time image linking.";
+                                String msg = String.format(msgFormat, file.getFile());
+                                IllegalArgumentException noModulePatch = new IllegalArgumentException(msg);
+                                throw new RuntimeImageLinkException(noModulePatch);
+                            }
+                        }
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                }
+
+                @Override
+                public InputStream stream() throws IOException {
+                    if (resType != EntryType.CLASS_OR_RESOURCE) {
+                        // Read from the base JDK image.
+                        Path path = symlink ? BASE.resolve(sha) : BASE.resolve(resPath);
+                        return Files.newInputStream(path);
+                    } else {
+                        // Read from the module image. Use the diff to the
+                        // packaged modules if we have one. Diffs of kind
+                        // ADDED have been filtered out in collectFiles();
+                        if (diff != null) {
+                            assert diff.getKind() != ResourceDiff.Kind.ADDED;
+                            assert diff.getName().equals(String.format("/%s/%s",
+                                                                    archive.moduleName(),
+                                                                    resPath));
+                            return new ByteArrayInputStream(diff.getResourceBytes());
+                        }
+                        String module = archive.moduleName();
+                        ModuleReference mRef = ModuleFinder.ofSystem()
+                                                    .find(module).orElseThrow();
+                        return mRef.open().open(resPath).orElseThrow();
+                    }
+                }
+
+            };
+        }
+    }
+
+    private static List<String> readModuleResourceFile(String modName) {
+        String resName = String.format(RESPATH_PATTERN, modName);
+        try {
+            try (InputStream inStream = JRTArchive.class.getModule()
+                                                  .getResourceAsStream(resName)) {
+                String input = new String(inStream.readAllBytes(),
+                                          StandardCharsets.UTF_8);
+                if (input.isEmpty()) {
+                    // Not all modules have non-class resources
+                    return Collections.emptyList();
+                } else {
+                    return Arrays.asList(input.split("\n"));
+                }
+            }
+        } catch (IOException e) {
+            throw new InternalError("Failed to process run-time image resources " +
```

Comment:
For unexpected exceptions, some code throws InternalError and some throws AssertionError - better to be consistent and I think implementation typically throws InternalError.     For IOException, I think `UncheckedIOException` is appropriate.

Reply:
I've changed it to AssertionError, but can change it to UncheckedIOException instead. Will take another pass on this issue tomorrow.

Reply:
Having looked at this again the patch now consistently uses `AssertionError` which should be fine? Please let me know if that's an issue.

Reply:
I prefer `UncheckedIOException` as it clearly shows related to I/O which could happen due to interrupted I/O  operations.  

`AssertionError` should never happen or a bug in the implementation.